docs/pai-job-protocol.yaml

```diff
@@ -0,0 +1,223 @@
+    commands:
+      - String, required
+
+# To handle that a component may interact with different component differently, user is encouraged to place the codes handling such difference in the "deployments" field,
+# e.g., a job may get input data through wget, hdfc -dfs cp, copy, or just directly read from remote storage. This logic can be placed here.
+# In summary, the deployments field is responsible to make sure the job to run properly in a deployment specific runtime environment.
+# One could have many deployments, but only one deployment can be activated at runtime by specifying in "defaults". User can choose the deployment and specify in "defaults" at submission time.
+deployments:
+  - name: String, required
+    taskRoles:
+      <name>: String, required # Should be in taskRoles
+        preCommands:
+          - String, required # execute before the taskRole's command
+        postCommands:
+          - String, required # execute after the taskRole's command
+
+
+defaults: # optional, default cluster specific settings
+  virtualCluster: String, optional
+  deployment: String, optional # Should reference to deployment defined in deployments
+
+
+---
+# OpenPAI Job Protocol YAML Example for a Distributed TensorFlow Job
+
+protocolVersion: 2
+name: tensorflow_cifar10
+type: job
+version: 1.0
+contributor: Alice
+description: image classification, cifar10 dataset, tensorflow, distributed training
+
+prerequisites:
+  - protocolVersion: 2
+    name: tf_example
+    type: dockerimage
+    version: latest
+    contributor: Alice
+    description: python3.5, tensorflow
+    auth:
+      username: 81f1fd6a-2844-4072-982d-62371fa37bd3
+      passward: user1
+      registryuri: openpai.azurecr.io
+    uri: openpai/pai.example.tensorflow
+  - protocolVersion: 2
+    name: tensorflow_cifar10_model
+    type: output
+    version: latest
+    contributor: Alice
+    description: cifar10 data output
+    uri: hdfs://10.151.40.179:9000/core/cifar10_model
+  - protocolVersion: 2
+    name: tensorflow_cnnbenchmarks
+    type: script
+    version: 84820935288cab696c9c2ac409cbd46a1f24723d
+    contributor: MaggieQi
+    description: tensorflow benchmarks
+    uri: https://github.com/MaggieQi/benchmarks
+  - protocolVersion: 2
+    name: cifar10
+    type: data
+    version: latest
+    contributor: Alice
+    description: cifar10 dataset, image classification
+    uri:
+      - https://www.cs.toronto.edu/~kriz/cifar-10-python.tar.gz
+
+parameters:
+  model: resnet20
+  batchsize: 32
+
+jobRetryCount: 1
+taskRoles:
+  worker:
+    instance: 1
+    completion:
+      minFailedInstances: 1
+      minSucceededInstances: 1
+    taskRetryCount: 0
+    dockerImage: tf_example
+    data: cifar10
+    output: tensorflow_cifar10_model
+    script: tensorflow_cnnbenchmarks
+    extraContainerOptions:
+      shmMB: 64
+    resourcePerInstance:
```

Comment:
>resourcePerInstance [](start = 4, length = 19)

resorucesPerInstance?

Reply:
The original one is `resourcePerInstance`, should we change to `resourcesPerInstance`?

Reply:
As it contains multiple children, maybe resourcesPerInstance is more grammer correct? @yang.fan@163.com can how do you think about it?

---
In reply to: [270776814](https://github.com/Microsoft/pai/pull/2260#discussion_r270776814) [](ancestors = 270776814)

Reply:
No need to change, Yarn also call it "resource".

---
In reply to: [270779984](https://github.com/Microsoft/pai/pull/2260#discussion_r270779984) [](ancestors = 270779984,270776814)